src/python_testing/matter_testing_infrastructure/matter/testing/matter_testing.py

```diff
@@ -1369,6 +1369,61 @@ def user_verify_push_av_stream(self, prompt_msg: str) -> bool:
             error_message='Push AV Stream validation failed'
         )
 
+    async def device_reboot(self, factory_reset: bool = False):
+        """Reboot the Device Under Test (DUT), optionally performing a factory reset.
+
+        This method handles device reboots in both CI and development environments (via run_python_test.py test runner script)
+        and also manual testing scenarios (via user input). It expires existing sessions to allow for controllers to reconnect
+        to the DUT after the reboot. If factory_reset is True, it will clear previously connected controllers by removing KVS file.
+
+        Args:
+            factory_reset: If True, signals the test runner to perform a factory reset
+                (by removing KVS file) by writing "reset" to the restart_flag_file. # Functionality implemented by Raul.
+                If False, performs a simple reboot. Default is False.
+
+        Returns:
+            None
+        """
+        # Check if restart flag file is available (indicates test runner supports app restart)
+        restart_flag_file = self.get_restart_flag_file()
+
+        if not restart_flag_file:
+            # No restart flag file: ask user to manually reboot
+            self.wait_for_user_input(prompt_msg="Reboot the DUT. Press Enter when ready.\n")
+
+            # After manual reboot, expire previous sessions so that we can re-establish connections
+            LOGGER.info("Expiring sessions after manual device reboot")
+            if hasattr(self, 'th1') and self.th1:
+                self.th1.ExpireSessions(self.dut_node_id)
+            if hasattr(self, 'th2') and self.th2:
+                self.th2.ExpireSessions(self.dut_node_id)
+            LOGGER.info("Manual device reboot completed")
+
+        else:
+            try:
+                # Create the restart flag file to signal the test runner
+                with open(restart_flag_file, "w") as f:
+                    if factory_reset:
+                        f.write("reset")
+                    else:
+                        f.write("restart")
+                    LOGGER.info("Created restart flag file to signal app restart")
+
+                # The test runner will automatically wait for the app-ready-pattern before continuing
+                # Waiting a bit longer for the app to be ready and stable after restart.
+                await asyncio.sleep(1)
+
+                # Expire sessions and re-establish connections
+                if hasattr(self, 'th1') and self.th1:
```

Comment:
suggest consolidating with above so you can use the fabric admin and do it in one place.